### PR TITLE
adding pagination for queries returning multiple items

### PR DESCRIPTION
**Pull request by muku314115**
## Description

Closes: #XXXX

---

### Author Checklist

*All items are required. Please add a note to the item if the item is not applicable and please add links to any
relevant follow-up issues.*

I have...

- [ ] included the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json)
  in the PR title
- [ ] added `!` to the type prefix if API or client breaking change
- [ ] targeted the correct branch (
  see [PR Targeting](https://github.com/AssetMantle/modules/blob/master/CONTRIBUTING.md#pr-targeting))
- [ ] provided a link to the relevant issue or specification
- [ ] followed the guidelines
  for [building modules](https://github.com/AssetMantle/modules/blob/master/docs/building-modules)
- [ ] included the necessary unit and
  integration [tests](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#testing)
- [ ] added a changelog entry to `CHANGELOG.md`
- [ ] included comments for [documenting Go code](https://blog.golang.org/godoc)
- [ ] updated the relevant documentation or specification
- [ ] reviewed "Files changed" and left comments if necessary
- [ ] confirmed all CI checks have passed

### Reviewers Checklist

*All items are required. Please add a note if the item is not applicable and please add your handle next to the items
reviewed if you only reviewed selected items.*

I have...

- [ ] confirmed the
  correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] confirmed `!` in the type prefix if API or client breaking change
- [ ] confirmed all author checklist items have been addressed
- [ ] reviewed state machine logic
- [ ] reviewed API design and naming
- [ ] reviewed documentation is accurate
- [ ] reviewed tests and test coverage
- [ ] manually tested (if applicable)

### CODE REVIEW

File: helpers/base/mapper.go [hunk 1]
```diff
@@ -75,6 +76,21 @@ func (mapper mapper) Iterate(context context.Context, partialKey helpers.Key, ac
...
+	_, err := sdkTypesQuery.Paginate(store, pageRequest, func(_, value []byte) error {
+		mappable := mapper.mappablePrototype()
+		CodecPrototype().MustUnmarshal(value, mappable)
+		if accumulator(mappable) {
+			return nil
+		}
+		return nil
+	})
+	if err != nil {
+		panic("Pagination failure: " + err.Error())
```
Comment (deepanshutr) [inline helpers/base/mapper.go:91]: correct err will be: "paginated query failure:"

File: x/assets/internal/queries/assets/queryRequest.proto [hunk 2]
```diff
@@ -0,0 +1,8 @@
+syntax = "proto3";
+package assets.queries.assets;
+
+import "cosmos/base/query/v1beta1/pagination.proto";
+
+message QueryRequest {
+  cosmos.base.query.v1beta1.PageRequest pagination = 2;
```
Comment (deepanshutr) [inline x/assets/internal/queries/assets/queryRequest.proto:7]: proto field name should ideally have been named page_request and numbered 1

File: x/assets/internal/queries/assets/request.go [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+}
+func queryRequestFromInterface(request helpers.QueryRequest) *QueryRequest {
+	switch value := request.(type) {
+	case *QueryRequest:
+		return value
+	default:
+		return &QueryRequest{}
+	}
+}
+func newQueryRequest(pagination *query.PageRequest) helpers.QueryRequest {
```
Comment (deepanshutr) [inline x/assets/internal/queries/assets/request.go:76]: input parameter should be renamed to pageRequest

File: x/assets/internal/queries/assets/response.go [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+
+import (
+	"errors"
+
+	"github.com/AssetMantle/modules/helpers"
+	"github.com/AssetMantle/modules/helpers/base"
+	"github.com/AssetMantle/modules/x/assets/internal/mappable"
+)
+
+// type queryResponse struct {
```
Comment (deepanshutr) [inline x/assets/internal/queries/assets/response.go:14]: can remove commented ref struct now

File: x/assets/internal/queries/assets/queryKeeper.go [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+	"github.com/AssetMantle/modules/x/assets/internal/key"
+)
+
+type queryKeeper struct {
+	mapper helpers.Mapper
+}
+
+var _ helpers.QueryKeeper = (*queryKeeper)(nil)
+
+func (queryKeeper queryKeeper) Enquire(context context.Context, queryRequest helpers.QueryRequest) helpers.QueryResponse {
```
Comment (deepanshutr) [inline x/assets/internal/queries/assets/queryKeeper.go:20]: requesting modification of queryKeeper interface Enquire method to return error, just like transaction keeper. error doesn't have to bundled in the reponse object, it's redundant

Comment (deepanshutr) [reply x/assets/internal/queries/assets/queryKeeper.go:20]: PL Replicate everywhere

File: x/assets/internal/queries/assets/queryResponse.proto [hunk 6]
```diff
@@ -0,0 +1,12 @@
+syntax = "proto3";
+package assets.queries.assets;
+
+import "gogoproto/gogo.proto";
+import "x/assets/internal/mappable/mappable.proto";
+
+message QueryResponse {
+  option (gogoproto.goproto_getters) = false;
+  bool success = 1;
+  string error = 2;
```
Comment (deepanshutr) [inline x/assets/internal/queries/assets/queryResponse.proto:10]: error and success are redundant can remove both

Comment (deepanshutr) [reply x/assets/internal/queries/assets/queryResponse.proto:10]: Requesting replication everywhere

File: x/assets/internal/queries/assets/response.go [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+	if err := base.CodecPrototype().UnmarshalJSON(bytes, queryResponse); err != nil {
+		return nil, err
+	}
+
+	return queryResponse, nil
+}
+func responsePrototype() helpers.QueryResponse {
+	return &QueryResponse{}
+}
+func newQueryResponse(collection helpers.Collection, error error) *QueryResponse {
```
Comment (deepanshutr) [inline x/assets/internal/queries/assets/response.go:41]: error in response reduntant, can remove

Comment (deepanshutr) [reply x/assets/internal/queries/assets/response.go:41]: Requesting replication everywhere

File: x/classifications/internal/queries/classifications/queryRequest.proto [hunk 8]
```diff
@@ -0,0 +1,8 @@
+syntax = "proto3";
+package classifications.queries.classifications;
+
+import "cosmos/base/query/v1beta1/pagination.proto";
+
+message QueryRequest {
+  cosmos.base.query.v1beta1.PageRequest pagination = 2;
```
Comment (deepanshutr) [inline x/classifications/internal/queries/classifications/queryRequest.proto:7]: !

File: x/classifications/internal/queries/classifications/queryResponse.proto [hunk 9]
```diff
@@ -0,0 +1,12 @@
+syntax = "proto3";
+package classifications.queries.classifications;
+
+import "gogoproto/gogo.proto";
+import "x/classifications/internal/mappable/mappable.proto";
+
+message QueryResponse {
+  option (gogoproto.goproto_getters) = false;
+  bool success = 1;
```
Comment (deepanshutr) [inline x/classifications/internal/queries/classifications/queryResponse.proto:9]: !

File: x/classifications/internal/queries/classifications/request.go [hunk 10]
```diff
@@ -0,0 +1,78 @@
...
+	"github.com/cosmos/cosmos-sdk/types/query"
+	"net/http"
+	"strconv"
+
+	"github.com/AssetMantle/modules/helpers"
+	"github.com/AssetMantle/modules/helpers/base"
+	"github.com/AssetMantle/modules/helpers/constants"
+)
+
+// type queryRequest struct {
```
Comment (deepanshutr) [inline x/classifications/internal/queries/classifications/request.go:18]: !

File: x/classifications/internal/queries/classifications/response.go [hunk 11]
```diff
@@ -0,0 +1,55 @@
...
+	}
+
+	return queryResponse, nil
+}
+func responsePrototype() helpers.QueryResponse {
+	return &QueryResponse{}
+}
+func newQueryResponse(collection helpers.Collection, error error) *QueryResponse {
+	list := mappable.MappablesFromInterface(collection.GetList())
+	if error != nil {
```
Comment (deepanshutr) [inline x/classifications/internal/queries/classifications/response.go:43]: !

File: x/identities/internal/queries/identities/queryKeeper.go [hunk 12]
```diff
@@ -0,0 +1,35 @@
...
+	"github.com/AssetMantle/modules/x/identities/internal/key"
+)
+
+type queryKeeper struct {
+	mapper helpers.Mapper
+}
+
+var _ helpers.QueryKeeper = (*queryKeeper)(nil)
+
+func (queryKeeper queryKeeper) Enquire(context context.Context, queryRequest helpers.QueryRequest) helpers.QueryResponse {
```
Comment (deepanshutr) [inline x/identities/internal/queries/identities/queryKeeper.go:20]: !

File: x/identities/internal/queries/identities/queryRequest.proto [hunk 13]
```diff
@@ -0,0 +1,8 @@
+syntax = "proto3";
+package identities.queries.identities;
+
+import "cosmos/base/query/v1beta1/pagination.proto";
+
+message QueryRequest {
+  cosmos.base.query.v1beta1.PageRequest pagination = 2;
```
Comment (deepanshutr) [inline x/identities/internal/queries/identities/queryRequest.proto:7]: !

File: x/identities/internal/queries/identities/queryResponse.proto [hunk 14]
```diff
@@ -0,0 +1,12 @@
+syntax = "proto3";
+package identities.queries.identities;
+
+import "gogoproto/gogo.proto";
+import "x/identities/internal/mappable/mappable.proto";
+
+message QueryResponse {
+  option (gogoproto.goproto_getters) = false;
+  bool success = 1;
```
Comment (deepanshutr) [inline x/identities/internal/queries/identities/queryResponse.proto:9]: !

File: x/identities/internal/queries/identities/request.go [hunk 15]
```diff
@@ -0,0 +1,81 @@
...
+	"github.com/cosmos/cosmos-sdk/types/query"
+	"net/http"
+	"strconv"
+
+	"github.com/AssetMantle/modules/helpers"
+	"github.com/AssetMantle/modules/helpers/base"
+	"github.com/AssetMantle/modules/helpers/constants"
+)
+
+// type queryRequest struct {
```
Comment (deepanshutr) [inline x/identities/internal/queries/identities/request.go:19]: !

File: x/identities/internal/queries/identities/request.go [hunk 16]
```diff
@@ -0,0 +1,81 @@
...
+}
+func (*QueryRequest) FromCLI(cliCommand helpers.CLICommand, _ client.Context) (helpers.QueryRequest, error) {
+	if offset, err := strconv.Atoi(cliCommand.ReadString(constants.Offset)); err != nil {
+		return &QueryRequest{}, err
+	} else {
+		return newQueryRequest(&query.PageRequest{Offset: uint64(offset)}), nil
+	}
+}
+func (*QueryRequest) FromHTTPRequest(httpRequest *http.Request) (helpers.QueryRequest, error) {
+	x := httpRequest.URL.Query()
```
Comment (deepanshutr) [inline x/identities/internal/queries/identities/request.go:47]: pl self review

File: x/identities/internal/queries/identities/response.go [hunk 17]
```diff
@@ -0,0 +1,55 @@
...
+
+import (
+	"errors"
+
+	"github.com/AssetMantle/modules/helpers"
+	"github.com/AssetMantle/modules/helpers/base"
+	"github.com/AssetMantle/modules/x/identities/internal/mappable"
+)
+
+// type queryResponse struct {
```
Comment (deepanshutr) [inline x/identities/internal/queries/identities/response.go:14]: !

File: x/identities/internal/queries/identities/response.go [hunk 18]
```diff
@@ -0,0 +1,55 @@
...
+	}
+
+	return queryResponse, nil
+}
+func responsePrototype() helpers.QueryResponse {
+	return &QueryResponse{}
+}
+func newQueryResponse(collection helpers.Collection, error error) *QueryResponse {
+	list := mappable.MappablesFromInterface(collection.GetList())
+	if error != nil {
```
Comment (deepanshutr) [inline x/identities/internal/queries/identities/response.go:43]: !

File: x/maintainers/internal/queries/maintainers/queryKeeper.go [hunk 19]
```diff
@@ -0,0 +1,35 @@
...
+	"github.com/AssetMantle/modules/x/maintainers/internal/key"
+)
+
+type queryKeeper struct {
+	mapper helpers.Mapper
+}
+
+var _ helpers.QueryKeeper = (*queryKeeper)(nil)
+
+func (queryKeeper queryKeeper) Enquire(context context.Context, queryRequest helpers.QueryRequest) helpers.QueryResponse {
```
Comment (deepanshutr) [inline x/maintainers/internal/queries/maintainers/queryKeeper.go:20]: !

File: x/maintainers/internal/queries/maintainers/queryRequest.proto [hunk 20]
```diff
@@ -0,0 +1,8 @@
+syntax = "proto3";
+package maintainers.queries.maintainers;
+
+import "cosmos/base/query/v1beta1/pagination.proto";
+
+message QueryRequest {
+  cosmos.base.query.v1beta1.PageRequest pagination = 2;
```
Comment (deepanshutr) [inline x/maintainers/internal/queries/maintainers/queryRequest.proto:7]: !

File: x/maintainers/internal/queries/maintainers/queryResponse.proto [hunk 21]
```diff
@@ -0,0 +1,12 @@
+syntax = "proto3";
+package maintainers.queries.maintainers;
+
+import "gogoproto/gogo.proto";
+import "x/maintainers/internal/mappable/mappable.proto";
+
+message QueryResponse {
+  option (gogoproto.goproto_getters) = false;
+  bool success = 1;
```
Comment (deepanshutr) [inline x/maintainers/internal/queries/maintainers/queryResponse.proto:9]: !

File: x/maintainers/internal/queries/maintainers/response.go [hunk 22]
```diff
@@ -0,0 +1,55 @@
...
+
+import (
+	"errors"
+
+	"github.com/AssetMantle/modules/helpers"
+	"github.com/AssetMantle/modules/helpers/base"
+	"github.com/AssetMantle/modules/x/maintainers/internal/mappable"
+)
+
+// type queryResponse struct {
```
Comment (deepanshutr) [inline x/maintainers/internal/queries/maintainers/response.go:14]: !

File: x/maintainers/internal/queries/maintainers/response.go [hunk 23]
```diff
@@ -0,0 +1,55 @@
...
+	}
+
+	return queryResponse, nil
+}
+func responsePrototype() helpers.QueryResponse {
+	return &QueryResponse{}
+}
+func newQueryResponse(collection helpers.Collection, error error) *QueryResponse {
+	list := mappable.MappablesFromInterface(collection.GetList())
+	if error != nil {
```
Comment (deepanshutr) [inline x/maintainers/internal/queries/maintainers/response.go:43]: !

File: x/metas/internal/queries/metas/queryRequest.proto [hunk 24]
```diff
@@ -0,0 +1,8 @@
+syntax = "proto3";
+package metas.queries.metas;
+
+import "cosmos/base/query/v1beta1/pagination.proto";
+
+message QueryRequest {
+  cosmos.base.query.v1beta1.PageRequest pagination = 2;
```
Comment (deepanshutr) [inline x/metas/internal/queries/metas/queryRequest.proto:7]: !

File: x/metas/internal/queries/metas/queryResponse.proto [hunk 25]
```diff
@@ -0,0 +1,12 @@
+syntax = "proto3";
+package metas.queries.metas;
+
+import "gogoproto/gogo.proto";
+import "x/metas/internal/mappable/mappable.proto";
+
+message QueryResponse {
+  option (gogoproto.goproto_getters) = false;
+  bool success = 1;
```
Comment (deepanshutr) [inline x/metas/internal/queries/metas/queryResponse.proto:9]: !

File: x/metas/internal/queries/metas/request.go [hunk 26]
```diff
@@ -0,0 +1,78 @@
...
+	"github.com/cosmos/cosmos-sdk/types/query"
+	"net/http"
+	"strconv"
+
+	"github.com/AssetMantle/modules/helpers"
+	"github.com/AssetMantle/modules/helpers/base"
+	"github.com/AssetMantle/modules/helpers/constants"
+)
+
+// type queryRequest struct {
```
Comment (deepanshutr) [inline x/metas/internal/queries/metas/request.go:18]: !

File: x/metas/internal/queries/metas/response.go [hunk 27]
```diff
@@ -0,0 +1,55 @@
...
+
+import (
+	"errors"
+
+	"github.com/AssetMantle/modules/helpers"
+	"github.com/AssetMantle/modules/helpers/base"
+	"github.com/AssetMantle/modules/x/metas/internal/mappable"
+)
+
+// type queryResponse struct {
```
Comment (deepanshutr) [inline x/metas/internal/queries/metas/response.go:14]: !

File: x/metas/internal/queries/metas/response.go [hunk 28]
```diff
@@ -0,0 +1,55 @@
...
+	}
+
+	return queryResponse, nil
+}
+func responsePrototype() helpers.QueryResponse {
+	return &QueryResponse{}
+}
+func newQueryResponse(collection helpers.Collection, error error) *QueryResponse {
+	list := mappable.MappablesFromInterface(collection.GetList())
+	if error != nil {
```
Comment (deepanshutr) [inline x/metas/internal/queries/metas/response.go:43]: !

File: x/orders/internal/queries/orders/queryKeeper.go [hunk 29]
```diff
@@ -0,0 +1,35 @@
...
+	"github.com/AssetMantle/modules/x/orders/internal/key"
+)
+
+type queryKeeper struct {
+	mapper helpers.Mapper
+}
+
+var _ helpers.QueryKeeper = (*queryKeeper)(nil)
+
+func (queryKeeper queryKeeper) Enquire(context context.Context, queryRequest helpers.QueryRequest) helpers.QueryResponse {
```
Comment (deepanshutr) [inline x/orders/internal/queries/orders/queryKeeper.go:20]: !

File: x/orders/internal/queries/orders/queryRequest.proto [hunk 30]
```diff
@@ -0,0 +1,8 @@
+syntax = "proto3";
+package orders.queries.orders;
+
+import "cosmos/base/query/v1beta1/pagination.proto";
+
+message QueryRequest {
+  cosmos.base.query.v1beta1.PageRequest pagination = 2;
```
Comment (deepanshutr) [inline x/orders/internal/queries/orders/queryRequest.proto:7]: !

File: x/orders/internal/queries/orders/queryResponse.proto [hunk 31]
```diff
@@ -0,0 +1,12 @@
+syntax = "proto3";
+package orders.queries.orders;
+
+import "gogoproto/gogo.proto";
+import "x/orders/internal/mappable/mappable.proto";
+
+message QueryResponse {
+  option (gogoproto.goproto_getters) = false;
+  bool success = 1;
```
Comment (deepanshutr) [inline x/orders/internal/queries/orders/queryResponse.proto:9]: !

File: x/orders/internal/queries/orders/request.go [hunk 32]
```diff
@@ -0,0 +1,78 @@
...
+	"github.com/cosmos/cosmos-sdk/types/query"
+	"net/http"
+	"strconv"
+
+	"github.com/AssetMantle/modules/helpers"
+	"github.com/AssetMantle/modules/helpers/base"
+	"github.com/AssetMantle/modules/helpers/constants"
+)
+
+// type queryRequest struct {
```
Comment (deepanshutr) [inline x/orders/internal/queries/orders/request.go:18]: !

File: x/orders/internal/queries/orders/response.go [hunk 33]
```diff
@@ -0,0 +1,55 @@
...
+	}
+
+	return queryResponse, nil
+}
+func responsePrototype() helpers.QueryResponse {
+	return &QueryResponse{}
+}
+func newQueryResponse(collection helpers.Collection, error error) *QueryResponse {
+	list := mappable.MappablesFromInterface(collection.GetList())
+	if error != nil {
```
Comment (deepanshutr) [inline x/orders/internal/queries/orders/response.go:43]: !

File: x/orders/internal/queries/orders/service.proto [hunk 34]
```diff
@@ -0,0 +1,12 @@
+syntax = "proto3";
+package orders.queries.orders;
+
+import "google/api/annotations.proto";
+import "x/orders/internal/queries/orders/queryRequest.proto";
+import "x/orders/internal/queries/orders/queryResponse.proto";
+
+service Service{
+  rpc Handle(QueryRequest) returns(QueryResponse){
+    option (google.api.http).get = "/mantle/orders/v1beta1/orders";
```
Comment (deepanshutr) [inline x/orders/internal/queries/orders/service.proto:10]: !

File: x/splits/internal/queries/splits/queryResponse.proto [hunk 35]
```diff
@@ -0,0 +1,12 @@
+syntax = "proto3";
+package splits.queries.splits;
+
+import "gogoproto/gogo.proto";
+import "x/splits/internal/mappable/mappable.proto";
+
+message QueryResponse {
+  option (gogoproto.goproto_getters) = false;
+  bool success = 1;
```
Comment (deepanshutr) [inline x/splits/internal/queries/splits/queryResponse.proto:9]: !

File: x/splits/internal/queries/splits/response.go [hunk 36]
```diff
@@ -0,0 +1,55 @@
...
+
+import (
+	"errors"
+
+	"github.com/AssetMantle/modules/helpers"
+	"github.com/AssetMantle/modules/helpers/base"
+	"github.com/AssetMantle/modules/x/splits/internal/mappable"
+)
+
+// type queryResponse struct {
```
Comment (deepanshutr) [inline x/splits/internal/queries/splits/response.go:14]: !

**Review by deepanshutr — CHANGES_REQUESTED** [review]
requesting commented changes

File: helpers/base/module.go [hunk 37]
```diff
@@ -73,7 +73,6 @@ func (module module) RegisterRESTRoutes(context client.Context, router *mux.Rout
 	router.HandleFunc("/"+module.Name()+"/parameters", module.parameterManagerPrototype().RESTQueryHandler(context)).Methods("GET")
 
 	for _, query := range module.queriesPrototype().GetList() {
-		router.HandleFunc("/"+module.Name()+"/"+query.GetName()+fmt.Sprintf("/{%s}", query.GetName()), query.RESTQueryHandler(context)).Methods("GET")
```
[the remark this answers is not in the document]
Comment (muku314115) [reply helpers/base/module.go:76]: eg. 
identitites/identity?idenity={id} or identities/identities?offset=0&limit=10

**Pull request merged**